src/protocol/requests/joinGroup/v4/response.js

```diff
@@ -0,0 +1,50 @@
+const Decoder = require('../../../decoder')
+const { failure, createErrorFromCode, failIfVersionNotSupported } = require('../../../error')
+
+/**
+ * JoinGroup Response (Version: 4) => throttle_time_ms error_code generation_id group_protocol leader_id member_id [members]
+ *   throttle_time_ms => INT32
+ *   error_code => INT16
+ *   generation_id => INT32
+ *   group_protocol => STRING
+ *   leader_id => STRING
+ *   member_id => STRING
+ *   members => member_id member_metadata
+ *     member_id => STRING
+ *     member_metadata => BYTES
+ */
+
+const decode = async rawData => {
+  const decoder = new Decoder(rawData)
+  const throttleTime = decoder.readInt32()
+  const errorCode = decoder.readInt16()
+
+  failIfVersionNotSupported(errorCode)
+
+  return {
+    throttleTime,
+    errorCode,
+    generationId: decoder.readInt32(),
+    groupProtocol: decoder.readString(),
+    leaderId: decoder.readString(),
+    memberId: decoder.readString(),
+    members: decoder.readArray(decoder => ({
+      memberId: decoder.readString(),
+      memberMetadata: decoder.readBytes(),
+    })),
+  }
+}
+
+const parse = async data => {
+  // KIP-394: Require member.id for initial join group request
+  if (data.errorCode !== 79 && failure(data.errorCode)) {
```

Comment:
Oh, I see now why it wasn't throwing in broker.joinGroup. This is pretty nasty in my opinion. We always throw protocol errors rather than returning with error codes, and calling code shouldn't have to interact with protocol error codes and stuff like that. It should just be protocol errors, so that you don't have to know magic numbers like 79.

Reply:
Yeah, i thought about that at first but at the moment if we throw error we will not propagate response with new member_id so we wont be able to use this one at the broker level, we can try to change parse function to add member id to the error response, what do you think about that ?